dbms/src/Common/SensitiveDataMasker.h

```diff
@@ -0,0 +1,101 @@
+
+#include <iostream>
+#include <memory>
+#include <string>
+#include <vector>
+#include <re2/re2.h>
+#include <re2/stringpiece.h>
+#include <Common/Exception.h>
+
+
+//* TODO: option with hyperscan? https://software.intel.com/en-us/articles/why-and-how-to-replace-pcre-with-hyperscan
+// re2::set should also work quite fast, but it doesn't return the match position, only which regexp was matched
+
+namespace DB
+{
+namespace ErrorCodes
+{
+    extern const int CANNOT_COMPILE_REGEXP;
+}
+
+
+class SensitiveDataMasker
+{
+private:
+    class MaskingRule
+    {
+    private:
+        const std::string name;
+        const std::string replacement_string;
+        const std::string regexp_string;
+
+        const RE2 regexp;
+        const re2::StringPiece replacement;
+
+        uint64_t matches_count = 0;
+        std::mutex mutex;
+
+        //options
+    public:
+        MaskingRule(const std::string & name, const std::string & regexpString, const std::string & replacementString)
+            : name(name)
+            , replacement_string(replacementString)
+            , regexp_string(regexpString)
+            , regexp(regexp_string, RE2::Quiet)
+            , replacement(replacement_string)
+        {
+            if (!regexp.ok())
+                throw DB::Exception(
+                    "SensitiveDataMasker: cannot compile re2: " + regexpString
+                        + ", error: " + regexp.error()
+                        + ". Look at https://github.com/google/re2/wiki/Syntax for reference.",
+                    DB::ErrorCodes::CANNOT_COMPILE_REGEXP);
+        }
+        int apply(std::string & data)
+        {
+            auto m = RE2::GlobalReplace(&data, regexp, replacement);
+            std::lock_guard lock(mutex);
+            matches_count += m;
+            return m;
+        }
+
+        const std::string & getName() const { return name; }
+        const std::string & getReplacementString() const { return replacement_string; }
+        uint64_t getMatchesCount() const { return matches_count; }
+    };
+
+    std::vector<std::unique_ptr<MaskingRule>> all_masking_rules;
+
+public:
+    SensitiveDataMasker() {}
+
+    void addMaskingRule(const std::string & name, const std::string & regexpString, const std::string & replacementString)
+    {
+        all_masking_rules.push_back(std::make_unique<MaskingRule>(name, regexpString, replacementString));
```

Comment:
тогда сдесь можно сделать `all_masking_rules.emplace_back(name, regexpString, replacementString);`

Reply:
Эх... Т.е. emplace не копирует. Я искал способ сразу положить объект в вектор без копирования, но оно почему-то никак не гуглится :) Поэтому такую кривобокую конструкцию с unique_ptr нагородил. Спасибо.

Reply:
Нет, похоже что тоже копирует.
```

/usr/bin/../lib/gcc/x86_64-linux-gnu/8/../../../../include/c++/8/bits/stl_construct.h:75:38: error: call to implicitly-deleted copy constructor of 'DB::SensitiveDataMasker::MaskingRule'
    { ::new(static_cast<void*>(__p)) _T1(std::forward<_Args>(__args)...); }
                                     ^   ~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/bin/../lib/gcc/x86_64-linux-gnu/8/../../../../include/c++/8/bits/stl_uninitialized.h:83:8: note: in instantiation of function template specialization 'std::_Construct<DB::SensitiveDataMasker::MaskingRule, DB::SensitiveDataMasker::MaskingRule>' requested here
                std::_Construct(std::__addressof(*__cur), *__first);
                     ^
/usr/bin/../lib/gcc/x86_64-linux-gnu/8/../../../../include/c++/8/bits/stl_uninitialized.h:134:2: note: in instantiation of function template specialization 'std::__uninitialized_copy<false>::__uninit_copy<std::move_iterator<DB::SensitiveDataMasker::MaskingRule *>, DB::SensitiveDataMasker::MaskingRule *>' requested here
        __uninit_copy(__first, __last, __result);
        ^
/usr/bin/../lib/gcc/x86_64-linux-gnu/8/../../../../include/c++/8/bits/stl_uninitialized.h:289:19: note: in instantiation of function template specialization 'std::uninitialized_copy<std::move_iterator<DB::SensitiveDataMasker::MaskingRule *>, DB::SensitiveDataMasker::MaskingRule *>' requested here
    { return std::uninitialized_copy(__first, __last, __result); }
                  ^
/usr/bin/../lib/gcc/x86_64-linux-gnu/8/../../../../include/c++/8/bits/stl_uninitialized.h:310:19: note: in instantiation of function template specialization 'std::__uninitialized_copy_a<std::move_iterator<DB::SensitiveDataMasker::MaskingRule *>, DB::SensitiveDataMasker::MaskingRule *, DB::SensitiveDataMasker::MaskingRule>' requested here
      return std::__uninitialized_copy_a
                  ^
/usr/bin/../lib/gcc/x86_64-linux-gnu/8/../../../../include/c++/8/bits/vector.tcc:446:13: note: in instantiation of function template specialization 'std::__uninitialized_move_if_noexcept_a<DB::SensitiveDataMasker::MaskingRule *, DB::SensitiveDataMasker::MaskingRule *, std::allocator<DB::SensitiveDataMasker::MaskingRule> >' requested here
            = std::__uninitialized_move_if_noexcept_a
                   ^
/usr/bin/../lib/gcc/x86_64-linux-gnu/8/../../../../include/c++/8/bits/vector.tcc:109:4: note: in instantiation of function template specialization 'std::vector<DB::SensitiveDataMasker::MaskingRule, std::allocator<DB::SensitiveDataMasker::MaskingRule> >::_M_realloc_insert<const std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > &, const std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > &, const std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > &>' requested here
          _M_realloc_insert(end(), std::forward<_Args>(__args)...);
          ^
/home/mfilimonov/workspace/ClickHouse/dbms/src/Common/SensitiveDataMasker.h:72:27: note: in instantiation of function template specialization 'std::vector<DB::SensitiveDataMasker::MaskingRule, std::allocator<DB::SensitiveDataMasker::MaskingRule> >::emplace_back<const std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > &, const std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > &, const std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > &>' requested here
        all_masking_rules.emplace_back(name, regexpString, replacementString);
                          ^
/home/mfilimonov/workspace/ClickHouse/dbms/src/Common/SensitiveDataMasker.h:32:19: note: copy constructor of 'MaskingRule' is implicitly deleted because field 'regexp' has a deleted copy constructor
        const RE2 regexp;
                  ^
/home/mfilimonov/workspace/ClickHouse/contrib/re2/re2/re2.h:751:3: note: 'RE2' has been explicitly marked deleted here
  RE2(const RE2&) = delete;
 ```

Можно как-то ещё это обойти?

Reply:
https://stackoverflow.com/questions/40457302/c-vector-emplace-back-calls-copy-constructor
Т.е. он не копирует, но требует чтобы хранящийся объект выполнял контракт CopyConstructible  на случай если потребуется увеличить вектор. Может просто оставлю как есть.